c/tskit/trees.c

```diff
@@ -7814,75 +7835,455 @@ remove_unextended(edge_list_t **head, edge_list_t **tail)
+                e_in = ex_in->edge;
+                if (e_in == old_edge) {
+                    self->near_side[e_in] = there;
+                    if (self->far_side[e_in] != there) {
+                        ex_in->extended = 1;
+                    }
+                    self->next_degree[child] -= 1;
+                    self->next_degree[self->parent_in[child]] -= 1;
+                    self->parent_in[child] = TSK_NULL;
+                }
+            }
+        }
+    }
+out:
+    return ret;
+}
+
+static float
+haplotype_extender_mergeable(haplotype_extender_t *self, tsk_id_t c)
+{
+    // returns the number of new edges needed
+    // if the paths in parent_in and parent_out
+    // up through nodes that aren't in the other tree
+    // end at the same place and don't have conflicting times;
+    // otherwise, return infinity
+    tsk_id_t p_in, p_out, child;
+    float num_new_edges; // needs to be float so we can have infinity
+    int num_extended;
+    double t_in, t_out;
+    bool climb_in, climb_out;
+
+    p_out = self->parent_out[c];
+    p_in = self->parent_in[c];
+    t_out = (p_out == TSK_NULL) ? INFINITY : self->ts->tables->nodes.time[p_out];
+    t_in = (p_in == TSK_NULL) ? INFINITY : self->ts->tables->nodes.time[p_in];
+    child = c;
+    num_new_edges = 0;
+    num_extended = 0;
+    while (true) {
+        climb_in = (p_in != TSK_NULL && self->last_degree[p_in] == 0
+                    && self->not_sample[p_in] && t_in < t_out);
+        climb_out = (p_out != TSK_NULL && self->next_degree[p_out] == 0
+                     && self->not_sample[p_out] && t_out < t_in);
+        if (climb_in) {
+            if (self->parent_in[child] != p_in) {
+                num_new_edges += 1;
+            }
+            child = p_in;
+            p_in = self->parent_in[p_in];
+            t_in = (p_in == TSK_NULL) ? INFINITY : self->ts->tables->nodes.time[p_in];
+        } else if (climb_out) {
+            if (self->parent_out[child] != p_out) {
+                num_new_edges += 1;
+            }
+            child = p_out;
+            p_out = self->parent_out[p_out];
+            t_out = (p_out == TSK_NULL) ? INFINITY : self->ts->tables->nodes.time[p_out];
+            num_extended += 1;
+        } else {
+            break;
+        }
     }
-    edges_in_head = NULL;
-    edges_in_tail = NULL;
-    edges_out_head = NULL;
-    edges_out_tail = NULL;
-    e_out = 0; // only to avoid an 'maybe uninitialized' compile warning
+    if ((num_extended == 0) || (p_in != p_out) || (p_in == TSK_NULL)) {
+        num_new_edges = INFINITY;
+    }
+    return num_new_edges;
+}
 
-    if (forwards) {
+static int
+haplotype_extender_merge_paths(
+    haplotype_extender_t *self, tsk_id_t c, double left, double right)
+{
+    int ret = 0;
+    tsk_id_t p_in, p_out, child;
+    double t_in, t_out;
+    bool climb_in, climb_out;
+
+    p_out = self->parent_out[c];
+    p_in = self->parent_in[c];
+    t_out = self->ts->tables->nodes.time[p_out];
```

Comment:
Same here - you don't want to be chasing 3 levels of pointers here in a loop. Grab a const pointer at the start of the function and the compiler will be able to do a *way* better job.